### PR TITLE
Define the `PollingExecutionContextScheduler` API

### PR DESCRIPTION
**Pull request by armanbilge**
This PR defines the API for a `PollingExecutionContextScheduler`, which offers methods to:
- `execute` a task, effectively immediately;
- `schedule` a callback to be invoked in the future; and
- `register` to be notified of read- and write-ready events on a file descriptor

### Expected implementors
- the libuv runtime in scala-native-loop
- the epoll runtime in epollcat
- the kqueue runtime in epollcat
- the io_uring runtime in fs2-io_uring
- the cURL runtime in http4s-curl

### Expected consumers
- snunit
- scala-native-http-client-async
- epollcat JDK NIO implementations

**If we do this right, then it means that any application using a runtime that implements this API would be fully interoperable with any library that consumes this API.**

I have invited the following collaborators to this project:
- @ekrich, representing Scala Native core
- @lolgab, representing scala-native-loop, snunit, and scala-native-http-client-async
- @sideeffffect, representing ZIO
- @keynmol, representing the goal to have "**one** cross-OS artifact that users can use which will do various epoll/kqueue/iouring things **under the hood**"

Of course, feedback is welcome from any and all onlookers as well. Particularly those that know better than me 😅 

Everything subject to question, debate, bikeshed, etc. Thanks!

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-ThisBuild / tlBaseVersion := "0.0"
+ThisBuild / tlBaseVersion := "1.0"
```
Comment (armanbilge) [inline build.sbt:1]: I'm dead serious about compatibility on this. No 0.0.x gang! Let's do our best to get it right 😇

**Comment by armanbilge** [comment]
@jd557 makes an excellent suggestion on [Discord](https://discord.com/channels/632150470000902164/635668881951686686/1023118123882262599):

> I wonder if instead of `ExecutionContext with Scheduler with Polling` it would make more sense to reuse the java APIs and have a `ExecutionContextExecutorService with ScheduledExecutorService with Polling`
> 
> It has a few more methods, but I think they can all be emulated in terms of schedule (and then some ECs could bring optimized implementations for those cases)

https://www.scala-lang.org/api/2.13.9/scala/concurrent/ExecutionContextExecutorService.html
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ScheduledExecutorService.html

Personally I am a big fan of delegating to existing, established APIs instead of making up our own, for reasons of binary-compatiblity and cross-building. Note that the `ScheduledExecutorService` API is somewhat more powerful than the `Scheduler` currently defined here and thus would be a bit more annoying to implement. But if its APIs are useful in downstream libraries then it makes sense to offer them (although speaking for the Typelevel stack, we have no use for them).

If we go this route, then we should add the `ScheduledExecutorService` interface and friends to Javalib in Scala Native core.

File: build.sbt (hunk 1 above)
Comment (lolgab) [reply build.sbt:1]: If we want to have long backward compatibility guarantees, I'm not sure open traits are a good idea.
You can't add a method to an open trait without breaking binary compatibility.
If you go from:
```scala
package foo

trait Trait {
  def bar: Int
}
```
to:
```scala
package foo

trait Trait {
  def bar: Int
  def foo: Int = defaultValue
}
```

You get this from MiMa:
```
Found 1 issue when checking against foo:before:1
 * interface foo.Foo does not have a correspondent in current version
   filter with: ProblemFilter.exclude[MissingClassProblem]("foo.Foo")
1 targets failed
after.mimaReportBinaryIssues Failed binary compatibility check! Found 1 potential problems
```

An API based on a class can instead be evolved in a binary backward compatible way:
```scala
class ScalaNativeRuntime private (
  _polling: Polling,
  _scheduler: Scheduler
) {
  def polling: Polling = _polling
  def scheduler: Scheduler = _scheduler

  private def copy(
    polling: Polling = this.polling,
    scheduler: Scheduler = this.scheduler
  ) = new ScalaNativeRuntime(polling, scheduler)

  def withPolling(polling: Polling): ScalaNativeRuntime =
    copy(polling = polling)
  def withScheduler(scheduler: Scheduler): ScalaNativeRuntime =
    copy(scheduler = scheduler)
}
```

You can add new fields as long as you provide a default (which can fail gracefully, for example).

Comment (lolgab) [reply build.sbt:1]: I had a doubt about binary compatibility with open traits. But I found that the problem exists only in Scala 2.11 (which we don't have to support). So it is fine :) Sorry

Comment (lolgab) [reply build.sbt:1]: I had a doubt about binary compatibility with open traits. But I found that the problem exists only in Scala 2.11 (which we don't have to support). So it is fine :) Sorry

Comment (armanbilge) [reply build.sbt:1]: > I had a doubt about binary compatibility with open traits.

Yes, in Scala 2.11 you cannot add new methods to `trait`s with default implementations without breaking binary-compatibility. I don't expect us to add a method with a default implementation.

> But I found that the problem exists only in Scala 2.11 (which we don't have to support)

I think we can keep 2.11 as long as it's easy. If it gets in the way of bincompat, then perhaps we can drop it (or just break bincompat for it).

**Comment by ekrich** [comment]
This is a list of things I could find that are related.

So here is the main loop if I am not mistaken. It just adds a `Runnable` to the queue and just runs them until they are all run. The runnable itself can add more items to the queue.
https://github.com/scala-native/scala-native/blob/main/nativelib/src/main/scala/scala/scalanative/runtime/ExecutionContext.scala

Then I guess there would be `Thread` which represents the main thread.
https://github.com/scala-native/scala-native/blob/main/javalib/src/main/scala/java/lang/Thread.scala

I seems that we have a `genMain` as well.
https://github.com/scala-native/scala-native/blob/main/tools/src/main/scala/scala/scalanative/codegen/Generate.scala

Other issues:
https://github.com/scala-native/scala-native/issues/1626

**Comment by armanbilge** [comment]
Thanks to everyone for their feedback on this proposal. Although it did not take off as a standalone project, I have integrated the ideas into https://github.com/typelevel/cats-effect/pull/3314, which exposes a general API for polling on file descriptors.

I will be archiving this repo shortly, but if anyone is interested to pick up this effort I would be very supportive of a fork :)

**Pull request closed without merging**